cmd/bucket-stats.go

```diff
@@ -88,29 +80,18 @@ func (brs *BucketReplicationStats) Empty() bool {
 }
 
 // Clone creates a new BucketReplicationStats copy
-func (brs BucketReplicationStats) Clone() BucketReplicationStats {
-	c := BucketReplicationStats{
-		Stats: make(map[string]*BucketReplicationStat, len(brs.Stats)),
-	}
-	// This is called only by replicationStats cache and already holds a read lock before calling Clone()
+func (brs BucketReplicationStats) Clone() (c BucketReplicationStats) {
+	// This is called only by replicationStats cache and already holds a
+	// read lock before calling Clone()
+
+	c = brs
+	// We need to copy the map, so we do not reference the one in `brs`.
+	c.Stats = make(map[string]*BucketReplicationStat, len(brs.Stats))
 	for arn, st := range brs.Stats {
-		c.Stats[arn] = &BucketReplicationStat{
-			FailedSize:     atomic.LoadInt64(&st.FailedSize),
-			ReplicatedSize: atomic.LoadInt64(&st.ReplicatedSize),
-			ReplicaSize:    atomic.LoadInt64(&st.ReplicaSize),
-			FailedCount:    atomic.LoadInt64(&st.FailedCount),
-			PendingSize:    atomic.LoadInt64(&st.PendingSize),
-			PendingCount:   atomic.LoadInt64(&st.PendingCount),
-			Latency:        st.Latency.clone(),
-		}
+		// make a copy of `*st`
+		s := *st
+		c.Stats[arn] = &s
 	}
-	// update total counts across targets
-	c.FailedSize = atomic.LoadInt64(&brs.FailedSize)
-	c.FailedCount = atomic.LoadInt64(&brs.FailedCount)
-	c.PendingCount = atomic.LoadInt64(&brs.PendingCount)
-	c.PendingSize = atomic.LoadInt64(&brs.PendingSize)
-	c.ReplicaSize = atomic.LoadInt64(&brs.ReplicaSize)
-	c.ReplicatedSize = atomic.LoadInt64(&brs.ReplicatedSize)
```

Comment:
the counts here also need to be maintained for compatibility with older mc clients - the replication metrics API changed with multisite replication to report by ARN. Prior to that mc was using above fields in lines 107-113

Reply:
@donatello ^^

Reply:
I don't believe any API will change due to this PR. It's just simplification. The values should be copied in the assignment. Not sure I understand your point.

Reply:
We don't need atomic.LoadInt here as there is lock already taken

Reply:
The data structure changed in between so whereas metrics are now ARN specific it wasn't before. So you need to preserve the code here to ensure that upon upgrade we do not lose metrics that were currently captured.

Reply:
The counts will be copied in the `c = brs` line - not sure how it is affected by any upgrade.

Reply:
Probably because these variables are 'atomically' updated, so as long as Go will not have further races on this for integer updates. Then a Clone() simply doing `v1 = v`  would be suffice.

Reply:
Whenever the counts are updated a write lock is taken.